src/samples/iio+network/iio-magnetometer-console-and-mqtt-publish.c

```diff
@@ -0,0 +1,908 @@
+    else
+        history_full = 1;
+
+    /* Record new sample and calculate the moving sum */
+    for (f = 0; f < filter->num_fields; f++) {
+        /** A field is going to be overwritten if history is full, so decrease the history sum */
+        if (history_full)
+            filter->history_sum[f] -=
+                filter->history[filter->history_index * filter->num_fields + f];
+
+        filter->history[filter->history_index * filter->num_fields + f] = *data[f];
+        filter->history_sum[f] += *data[f];
+
+        /* For now simply compute a mobile mean for each field and output filtered data */
+        *data[f] = filter->history_sum[f] / filter->history_entries;
+    }
+
+    /* Update our rolling index (next evicted cell) */
+    filter->history_index = (filter->history_index + 1) % filter->history_size;
+}
+#endif
+
+static void
+iio_magnetometer_reader_cb(void *data, struct sol_iio_device *device)
+{
+    int r;
+    char buffer[128];
+    struct iio_magnetometer_data *magn_data = data;
+    struct sol_direction_vector out;
+    double azimuth;
+    struct sol_buffer payload;
+    struct sol_mqtt_message mqtt_message;
+
+#ifdef MAGN_CALIBRATE
+    struct compass_cal *cal_data = &(magn_data->cal_data);
+#endif
+#ifdef DENOISE_AVERAGE
+    struct filter_average *filter = &(magn_data->filter);
+#endif
+    r = sol_iio_read_channel_value(magn_data->channel_x, &out.x);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+    r = sol_iio_read_channel_value(magn_data->channel_y, &out.y);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+    r = sol_iio_read_channel_value(magn_data->channel_z, &out.z);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+#ifdef MAGN_CALIBRATE
+    out.x = CONVERT_GAUSS_TO_MICROTESLA(out.x);
+    out.y = CONVERT_GAUSS_TO_MICROTESLA(out.y);
+    out.z = CONVERT_GAUSS_TO_MICROTESLA(out.z);
+
+    sol_iio_mount_calibration(device, &out);
+
+    calibrate_compass(cal_data, &out.x, &out.y, &out.z);
+#endif
+
+#ifdef DENOISE_AVERAGE
+    denoise_average(magn_data->sampling_frequency, filter, &out.x, &out.y, &out.z);
+#endif
+
+    if ((out.x == 0) && (out.y == 0)) {
+        SOL_ERR("Point (0, 0) is invalid!\n");
+        return;
+    }
+
+    if (out.x == 0) {
+        if (out.y > 0)
+            azimuth = 0;
+        else
+            azimuth = 180;
+    } else if (out.y == 0) {
+        if (out.x > 0)
+            azimuth = 90;
+        else
+            azimuth = 270;
+    } else {
+        if (out.x > 0)
+            azimuth = 90 - atan(out.y / out.x) * 180 / M_PI;
+        else
+            azimuth = 270 - atan(out.y / out.x) * 180 / M_PI;
+    }
+
+#ifdef MAGN_CALIBRATE
+    snprintf(buffer, sizeof(buffer) - 1, "%d\t%d\t%d\t[uT]\t\t(Azimuth:%d)(Calibrated Level:%d)"
+        , (int)out.x, (int)out.y, (int)out.z, (int)(360 - azimuth), cal_data->cal_level);
+#else
+    snprintf(buffer, sizeof(buffer) - 1, "%d\t%d\t%d\t[uT]\t(Azimuth:%d)\t(Calibration disabled)"
+        , (int)out.x, (int)out.y, (int)out.z, (int)(360 - azimuth));
```

Comment:
why all those casts to `int` ? Wouldn't be better print the doubles with some specific precision ?